chalice/deploy/packager.py

```diff
@@ -208,3 +216,373 @@ def _add_chalice_lib_if_needed(self, project_dir, zip_fileobj):
+        contents = self._osutils.get_directory_contents(unpack_dir)
+        assert len(contents) == 1
+        return self._osutils.joinpath(unpack_dir, contents[0])
+
+    def get_package_name_and_version(self, directory, filename):
+        # type: (str, str) -> Tuple[str, str]
+        sdist_path = self._osutils.joinpath(directory, filename)
+        with self._osutils.tempdir() as tempdir:
+            package_dir = self._unpack_sdist_into_dir(sdist_path, tempdir)
+            pkg_info_filepath = self._generate_egg_info(package_dir)
+            metadata = self._parse_pkg_info_file(pkg_info_filepath)
+            name = metadata['Name']
+            version = metadata['Version']
+        return name, version
+
+
+class PipWrapper(object):
+    def main(self, args):
+        # type: (List[str]) -> Tuple[Optional[bytes], Optional[bytes]]
+        raise NotImplementedError('PipWrapper.main')
+
+
+class SubprocessPip(PipWrapper):
+    """Wrapper around calling pip through a subprocess."""
+    def main(self, args):
+        # type: (List[str]) -> Tuple[Optional[bytes], Optional[bytes]]
+        python_exe = sys.executable
+        invoke_pip = [python_exe, '-m', 'pip']
+        invoke_pip.extend(args)
+        p = subprocess.Popen(invoke_pip,
+                             stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out, err = p.communicate()
+        return out, err
+
+
+class PipRunner(object):
+    """Wrapper around pip calls used by chalice."""
+
+    def __init__(self, pip):
+        # type: (PipWrapper) -> None
+        self._wrapped_pip = pip
+
+    def _execute(self, command, args):
+        # type: (str, List[str]) -> None
+        """Execute a pip command with the given arguments."""
+        main_args = [command] + args
+        _, err = self._wrapped_pip.main(main_args)
+        if err:
+            if b'ReadTimeoutError' in err:
+                raise PipError('Read time out downloading dependencies.')
+            if b'NewConnectionError' in err:
+                raise PipError('Failed to establish a new connection when '
+                               'downloading dependencies.')
+            if b'PermissionError' in err:
+                match = re.search("Permission denied: '(.+?)'", str(err))
+                raise PipError('Do not have permissions to write to %s.'
+                               % match.group(1))
+
+    def build_wheel(self, wheel, directory):
+        # type: (str, str) -> None
+        """Build an sdist into a wheel file."""
+        arguments = ['--no-deps', '--wheel-dir', directory, wheel]
+        self._execute('wheel', arguments)
+
+    def download_all_dependencies(self, requirements_file, directory):
+        # type: (str, str) -> None
+        """Download all dependencies as sdist or whl."""
+        arguments = ['-r', requirements_file, '--dest', directory]
+        self._execute('download', arguments)
+
+    def download_manylinux_whls(self, packages, directory):
```

Comment:
Nitpick, but it's not 100% consistent in this code if we use "wheel" or "whl" (`build_wheel` vs. `download_manylinux_whls`.  FWIW, I prefer "wheel", it's two extra characters, makes it more readable, and is the name of the concept, whereas "whl" is the file extension.